js/binance.js

```diff
@@ -979,6 +989,110 @@ module.exports = class binance extends Exchange {
+        // token leftovers (of any asset) into `BNB` coin which in turn can be used to pay trading fees with it.
+        // The corresponding trades history is called the `Dust Log` and can be requested via the following end-point:
+        // https://github.com/binance-exchange/binance-official-api-docs/blob/master/wapi-api.md#dustlog-user_data
+        //
+        let request = this.extend ({}, params);
+        let response = await this.wapiGetUserAssetDribbletLog (request);
+        // { success:    true,
+        //   results: { total:    1,
+        //               rows: [ {     transfered_total: "1.06468458",
+        //                         service_charge_total: "0.02172826",
+        //                                      tran_id: 2701371634,
+        //                                         logs: [ {              tranId:  2701371634,
+        //                                                   serviceChargeAmount: "0.00012819",
+        //                                                                   uid: "35103861",
+        //                                                                amount: "0.8012",
+        //                                                           operateTime: "2018-10-07 17:56:07",
+        //                                                      transferedAmount: "0.00628141",
+        //                                                             fromAsset: "ADA"                  } ],
+        //                                 operate_time: "2018-10-07 17:56:06"                                } ] } }
+        let rows = response['results']['rows'];
+        let data = [];
+        for (let i = 0; i < rows.length; i++) {
+            let logs = rows[i]['logs'];
+            for (let j = 0; j < logs.length; j++) {
+                logs[j]['isDustTrade'] = true;
+                data.push (logs[j]);
+            }
+        }
+        const trades = this.parseTrades (data, undefined, since, limit);
+        return this.filterBySinceLimit (trades, since, limit);
+    }
+
+    parseDustTrade (trade, market = undefined) {
+        // {              tranId:  2701371634,
+        //   serviceChargeAmount: "0.00012819",
+        //                   uid: "35103861",
+        //                amount: "0.8012",
+        //           operateTime: "2018-10-07 17:56:07",
+        //      transferedAmount: "0.00628141",
+        //             fromAsset: "ADA"                  },
+        let order = this.safeString (trade, 'tranId');
+        let time = this.safeString (trade, 'operateTime');
+        let timestamp = this.parse8601 (time.replace (' ', 'T') + 'Z');
+        let datetime = this.iso8601 (timestamp);
+        let tradedCurrency = this.safeCurrencyCode (trade, 'fromAsset');
+        let earnedCurrency = this.currency ('BNB')['code'];
+        let tradedCurrencyIsQuote = false;
+        for (let i = 0; i < this.options.quoteCurrencyIds.length; i++) {
```

Comment:
This loop has to follow single-quoted assoc keys rule, because without quotes everything inside gets transpiled to properties, which are treated differently in Python and PHP.

Another concern on this – their dust trades are always X/BNB (that is, at least on my account), so, the earned currency should always be the quote, no?

Reply:
On the quote currency ids – I suggest that we don't add a mapping of quotes, but rather add one single option for the base – because it is always BNB. I mean, instead of adding exclusive mappings for every quote currency, we'd rather just designate the base in a single option? @den does that sound ok to you?

Reply:
> On the quote currency ids – I suggest that we don't add a mapping of quotes, but rather add one single option for the base and compare to it – because it is always BNB. I mean, instead of adding exclusive mappings for every quote currency, we'd rather just designate the base in a single option? @den does that sound ok to you? Otherwise we will have to list all currencies there again (it's hard to maintain a complete list of 300+ dust currencies and update it frequently).

Do you mean that we should assume that `BNB` is always a base currency? If so then i don't think it is possible, because (for reference) there's no `BNB/ADA` market at binance, there is only `ADA/BNB` one. So if we want to conform to actual binance markets we should define base and quote currencies accordingly.

Reply:
I don't like hardcoded quote currencies either so probably we could somehow rely on `this.markets_by_id` instead. Let me think of it a bit...

Reply:
> Do you mean that we should assume that BNB is always a base currency? 

No, I mean, that we should assume, that in a dust trade BNB is always the currency we get. And, based on that, we should look up either a `X/BNB` symbol in the `.markets` or `BNB/X`, whatever is present there – that lookup is linear, therefore it is very cheap and affordable. The main reasoning behind this is not to duplicate the list of currencies, because you've listed just 5 of 300. If I'd ask you to list the rest of them – you would have difficulties maintaining that list, because by the time you finish it – they will delist and add a couple of new items there. Thus, in general, maintaining any hardcoded mapping is a real pain in the long run, and we should not introduce such new mappings as much as possible. Does that make sense?

Reply:
```JavaScript
const earnedCurrency = this.currency ('BNB');
const symbol1 = earnedCurrency + '/' + otherCurrency;
const symbol2 = otherCurrency + '/' + earnedCurrency;
const market = this.safeValue2 (this.markets, earnedCurrency, otherCurrency);
if (market !== undefined) {
    symbol = market['symbol'];
}
```